### PR TITLE
Web3Single implementation

### LINKED ISSUE #12: Implement almost all of Web3Single
The `Web3Single` class needs to be implemented. 

https://github.com/NickChapman/RequestNetwork.py/blob/476897be1419f774f5ba245cb51f2e1105e5d658/servicesExternal/web3_single.py#L16

### PR DESCRIPTION
**Pull request by ghost**
Implemented most of `Web3Single`, closes #12

### CODE REVIEW

**Review by varuna82 — COMMENTED** [review]
@justicesuh 
Hi, This is just for my clarification, how does "send" function work? According to web3.py documentation "transact" function should be used to generate a transaction.
 (http://web3py.readthedocs.io/en/stable/contracts.html#web3.contract.Contract.transact)
```
method.send(options)
   .on('transactionHash', callbackTranactionHash)
   .on('receipt', callbackTransactionReceipt)
   .on('confirmation', callbackTransactionConfirmation)
   .on('error', callbackTransactionError)
```

**Comment by ghost** [comment]
@varuna82 You are correct, it should be `transact` instead of `send`

**Comment by varuna82** [comment]
@justicesuh 
I am bit skeptic about following lines too. I have tried to get following events in my test env, but it generates a runtime error saying "method not found". It looks like 'transactionHash' isn't supported by web3.py. Actually, web3.py and web3.js are totally different when it comes to event filtering.

@NickChapman what are your thoughts on this?

```
   .on('transactionHash', callbackTranactionHash)
   .on('receipt', callbackTransactionReceipt)
```

This could be the most difficult part to port. So, good luck with it...

**Comment by ghost** [comment]
@varuna82 Yea, web3.py and web3.js have very different apis (the web3.py api differs pretty significantly between versions too).  My thoughts were to port the general structure of the code first and then deal with differences in underlying api calls later (mainly web3 calls).

@NickChapman thoughts?

**Comment by NickChapman** [comment]
I would say open an issue and see if you can get it to raise an error for those events and we'll deal with it later. Also you'll need to resolve this merge conflict.

**Comment by justicesuh** [comment]
@NickChapman resolved

File: servicesExternal/web3_single.py [hunk 1]
```diff
@@ -42,58 +44,201 @@ def getNetworkName(networkId: int) -> str:
             42: 'kovan'
         }.get(networkId, 'private')
 
-    # Async
-    def broadcastMethod(self,
+    async def broadcastMethod(self,
                         method: Any,
                         callbackTranactionHash,
```
Comment (NickChapman) [inline servicesExternal/web3_single.py]: Can you annotate these callbacks with the `callable` type from `typing` https://docs.python.org/3/library/typing.html#callable

File: servicesExternal/web3_single.py [hunk 2]
```diff
@@ -42,58 +44,201 @@ def getNetworkName(networkId: int) -> str:
...
-
-    # Async
     def getDefaultAccountCallback(self, callback) -> None:
-        pass
+        '''
+        Get the default account (account[0] of the wallet) with a callback
+        '''
+        try:
+            accs = self.web3.eth.getAccounts()
+            if len(accs) is 0:
```
Comment (NickChapman) [inline servicesExternal/web3_single.py:136]: I wouldn't use `is` here, it works but I'm not sure if it is always guaranteed to work.

Comment (NickChapman) [reply servicesExternal/web3_single.py:136]: And actually you can just use `if accs`

File: servicesExternal/web3_single.py [hunk 3]
```diff
@@ -42,58 +44,201 @@ def getNetworkName(networkId: int) -> str:
...
+            return callback(e, None)
 
     def toSolidityBytes32(self, type: str, value) -> Any:
-        pass
+        '''
+        Convert a value in solidity bytes32 string
+        :param type: type of the value to convert (e.g: address, unint, int, etc...)
+        :param value: value to convert
+        '''
+        return self.web3.utils.encoding.to_hex(toSolidityBytes32(type, value))
```
Comment (NickChapman) [inline servicesExternal/web3_single.py:148]: recursively calls itself?

File: servicesExternal/web3_single.py [hunk 4]
```diff
@@ -42,58 +44,201 @@ def getNetworkName(networkId: int) -> str:
...
 
     def arrayToBytes32(self, array, length: int) -> List[Any]:
-        pass
+        '''
+        Convert an array to an array in solidity bytes32 string
+        TODO: only support addresses so far
```
Comment (NickChapman) [inline servicesExternal/web3_single.py:153]: Can you open an issue for this TODO

File: servicesExternal/web3_single.py [hunk 5]
```diff
@@ -42,58 +44,201 @@ def getNetworkName(networkId: int) -> str:
...
+        :param array: array to convert
+        :param length: length of the final array
+        '''
+        array = arrray or []
+        ret = []
+        for o in array:
+            ret.append(self.web3.utils.encoding.to_hex(toSolidityBytes32('address', 0)))
```
Comment (NickChapman) [inline servicesExternal/web3_single.py:160]: This is what `toSolidityBytes32` seems to do right above this?

File: servicesExternal/web3_single.py [hunk 6]
```diff
@@ -42,58 +44,201 @@ def getNetworkName(networkId: int) -> str:
...
+        # fill the empty case with zeroes
+        for i in range(len(array), length):
+            ret.append(self.web3.utils.encoding.to_hex(toSolidityBytes32'bytes32', 0))
```
Comment (NickChapman) [inline servicesExternal/web3_single.py]: Missing a `(` here

File: servicesExternal/web3_single.py [hunk 7]
```diff
@@ -42,58 +44,201 @@ def getNetworkName(networkId: int) -> str:
...
 
     def isAddressNoChecksum(self, address: str) -> bool:
-        if not address:
+        '''
+        Check if an address is valid (ignoring case)
+        :param address: address to check
+        '''
+        if not address :
             return False
         return address and self.web3.utils.isAddress(address.toLowerCase())
```
Comment (NickChapman) [inline servicesExternal/web3_single.py]: This isn't your code but looking at it now there's no need for the `address and`, it should just be `return self.....isAddress()`

File: servicesExternal/web3_single.py [hunk 8]
```diff
@@ -42,58 +44,201 @@ def getNetworkName(networkId: int) -> str:
...
+        :param address1: address to check
+        :param address2: address to check
+        '''
+        if not address1 or not address2 :
             return False
         return address1.toLowerCase() == address2.toLowerCase()
 
     def isHexStrictBytes32(self, hex: str) -> bool:
+        '''
+        Check if a string is a hex byte
```
Comment (NickChapman) [inline servicesExternal/web3_single.py:187]: `hex byte` but it's 66 hex characters so that can't be quite right

**Review by NickChapman — CHANGES_REQUESTED** [review]

File: servicesExternal/web3_single.py (hunk 3 above)
Comment (ghost) [reply servicesExternal/web3_single.py:148]: there's an imported function with the same name, will fix

File: servicesExternal/web3_single.py (hunk 4 above)
Comment (ghost) [reply servicesExternal/web3_single.py:153]: this was a TODO within the existing implementation of RequestNetwork.js

File: servicesExternal/web3_single.py (hunk 5 above)
Comment (ghost) [reply servicesExternal/web3_single.py:160]: I believe so?  I think the end goal was to have this method handle the different types of arrays

File: servicesExternal/web3_single.py (hunk 8 above)
Comment (ghost) [reply servicesExternal/web3_single.py:187]: It's '0x' + 32 * 2

**Review by NickChapman — APPROVED** [review]

**Pull request merged**